Cache result of slow function resolution on Any that is present on a hot path

hubertp opened the pull request:
### Pull Request Description

Otherwise things can go horribly slow.
Closes #6523. Follow up on https://github.com/enso-org/enso/pull/6441.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible. (Benchmarks now run as before)

hubertp committed:
Cache resolution of symbol for Any

Otherwise things can go horribly slow.
Closes #6523.

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -112,7 +112,8 @@ Object doFunctionalDispatch(
       @CachedLibrary(limit = "10") TypesLibrary typesLibrary,
-      @Cached MethodResolverNode methodResolverNode) {
+      @Cached MethodResolverNode methodResolverNode,
+      @Cached("anyFunction(symbol)") Function anyFun) {
```
I think that you should also cache `symbol` here, and check whether it is still same in the guards. In this form, the `anyFunc` is resolved during the *specialization instantiation*, but you can potentially get a different symbol every time.

My suggestion is: Add `@Cached("symbol") UnresolvedSymbol cachedSymbol` to the parameters, and `"cachedSymbol.equals(symbol)"` (or `"cachedSymbol == symbol"`) to the guards.

hubertp committed:
cache unresolvedsymbol

As per PR review

hubertp committed:
missing any condition in uncached

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -104,46 +104,73 @@ public void setTailStatus(TailStatus tailStatus) {
   }
 
+  @Specialization(replaces = "doFunctionalDispatchCachedSymbol", guards = {"typesLibrary.hasType(self)", "!typesLibrary.hasSpecialDispatch(self)"})
+  Object doFunctionalDispatchUncachedSymbol(
```
Is this really a 1:1 copy of `doFunctionalDispatchCachedSymbol`, just without caching the `anyFunction` resolution? Why is this necessary? What is the goal?

For each `UnresolvedSymbol symbol` there is only one query to `anyFunction(symbol)` (in a single `EnsoContext`) needed. It is possible that `cachedSymbol != symbol` (certaily in multiple contexts), but how can that happen in a single context? Does that happen? When?

In any case, the solution for multiple symbols is to built a polymorphic cache of certain depth/limit - e.g. use `limit=".."` in the previous `@Specialization`.

Then there has to be a "fallback" case when the limit is exceeded. Is this `@Specialization` supposed to be a fallback?

Akirathan replied:
I guess that this was needed, because `ModuleScope.lookupMethodDefinition` is an expensive method?

> Is it possible that cachedSymbol != symbol?

Is that even possible in multiple context? I mean, Enso method calls are "statically resolvable", are they not? Which means that for every call site, there is a fixed symbol. Am I missing something?

hubertp replied:
> I guess that this was needed, because `ModuleScope.lookupMethodDefinition` is an expensive method

Yes, it is pretty expensive especially on `Any`. The reason why Benchmarks were timing out.

> It is possible that `cachedSymbol != symbol`

I was getting a lot of failures on `Value_Type`'s tests without this condition. Also `invokeMember` wouldn't play nice, as indicated by failures in `TypeMembersTest`.

hubertp replied:
I think I addressed the comments in the latest change. It's good feedback btw, thank you.

JaroslavTulach replied:
`replaces = "doFunctionalDispatchCachedSymbol",` seems reasonable.

JaroslavTulach reviewed (approved):
I am not sure what is the goal of the change? I am surprised to see two copies of the same specialization. I don't understand the motivation for such change.

hubertp committed:
Rewrite caching of the resolved function

Applying PR review.

Had to be a bit careful so that we don't trip on automatic resource
managment on `self`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -104,30 +105,46 @@ public void setTailStatus(TailStatus tailStatus) {
+          "cachedSymbol == symbol",
+          "cachedSelfTpe == typesLibrary.getType(self)",
+          "function != null"
+  }, limit = "CACHE_SIZE")
```
Limit is good.
